README
$Id: README,v 1.7 2002/06/29 09:38:09 wiz Exp $

These services were originally written to work with ircnn-2.9.6 on newnet.


ChangeLog
wiz         2002/10/10 18:19:19 HST	(20021011_0)

  Modified files:
    include              net.h 
    src                  clone.c dcc.c net.c 
  Log:
  - Rewrote network parser to use stdio functions
  
  Revision  Changes    Path
  1.5       +2 -1      zorak-services/include/net.h
  1.16      +7 -7      zorak-services/src/clone.c
  1.5       +4 -2      zorak-services/src/dcc.c
  1.38      +33 -42    zorak-services/src/net.c



wiz         2002/10/02 13:28:51 HST	(20021002_7)

  Modified files:
    src                  main.c net.c 
  Log:
  - Fixed dcc cleanup code some more
  
  Revision  Changes    Path
  1.41      +1 -1      zorak-services/src/main.c
  1.37      +12 -2     zorak-services/src/net.c



wiz         2002/10/02 12:08:42 HST	(20021002_6)

  Modified files:
    src                  conf.c dcc.c net.c 
  Log:
  - Added M: line IP field for vhosting...
  - Fixed more DCC stuff and dcc cleanup
  
  Revision  Changes    Path
  1.22      +14 -12    zorak-services/src/conf.c
  1.4       +16 -1     zorak-services/src/dcc.c
  1.36      +4 -2      zorak-services/src/net.c



wiz         2002/10/02 12:08:42 HST	(20021002_5)

  Modified files:
    doc                  example.conf 
  Log:
  - Added M: line IP field for vhosting...
  
  Revision  Changes    Path
  1.8       +6 -6      zorak-services/doc/example.conf



wiz         2002/10/02 07:39:53 HST	(20021002_4)

  Modified files:
    src                  .depend conf.c dcc.c main.c net.c parse.c 
                         sclients.c users.c 
  Log:
  - Updated .depend
  - Added parsing for N0 client
  - Added client into intro_nick() loop
  - Added dcc_cleanup() and an add_event() for it in main()
  - Fixed gline() for real this time
  - Added entire DCC help system (still buggy and incomplete)
  
  Revision  Changes    Path
  1.8       +69 -40    zorak-services/src/.depend
  1.21      +2 -2      zorak-services/src/conf.c
  1.3       +158 -18   zorak-services/src/dcc.c
  1.40      +3 -1      zorak-services/src/main.c
  1.35      +16 -0     zorak-services/src/net.c
  1.58      +1 -0      zorak-services/src/parse.c
  1.2       +3 -3      zorak-services/src/sclients.c
  1.35      +6 -2      zorak-services/src/users.c



wiz         2002/10/02 07:39:53 HST	(20021002_3)

  Modified files:
    include              dcc.h defines.h 
  Log:
  - Added dcc_cleanup() and HELP_DESC, HELP_USAGE
  
  Revision  Changes    Path
  1.3       +1 -0      zorak-services/include/dcc.h
  1.3       +4 -0      zorak-services/include/defines.h



wiz         2002/10/01 18:57:28 HST	(20021002_2)

  Modified files:
    src                  users.c 
  Log:
  - Oops had an extra cptr->nick in the last commit
  
  Revision  Changes    Path
  1.34      +1 -1      zorak-services/src/users.c



wiz         2002/10/01 18:28:06 HST	(20021002_1)

  Modified files:
    src                  users.c 
  Log:
  - hyb7 gline fix for Jeff
  
  Revision  Changes    Path
  1.33      +15 -22    zorak-services/src/users.c



wiz         2002/10/01 16:56:20 HST	(20021002_0)

  Modified files:
    include              chanserv.h client.h conf.h dcc.h events.h 
                         flags.h irc_string.h me.h msg.h net.h 
                         nicks.h nickserv.h operserv.h parse.h 
    src                  chanserv.c clone.c conf.c dcc.c events.c 
                         irc_string.c main.c mem.c net.c 
                         nickserv.c oline.c operserv.c parse.c 
                         send.c users.c 
  Log:
  - First commit after moving cvs repo
  
  Revision  Changes    Path
  1.2       +2 -2      zorak-services/include/chanserv.h
  1.2       +0 -2      zorak-services/include/client.h
  1.2       +1 -0      zorak-services/include/conf.h
  1.2       +1 -0      zorak-services/include/dcc.h
  1.2       +1 -0      zorak-services/include/events.h
  1.6       +1 -1      zorak-services/include/flags.h
  1.3       +1 -0      zorak-services/include/irc_string.h
  1.3       +1 -0      zorak-services/include/me.h
  1.35      +35 -51    zorak-services/include/msg.h
  1.4       +6 -3      zorak-services/include/net.h
  1.2       +2 -0      zorak-services/include/nicks.h
  1.2       +5 -5      zorak-services/include/nickserv.h
  1.3       +18 -34    zorak-services/include/operserv.h
  1.2       +25 -27    zorak-services/include/parse.h
  1.6       +4 -8      zorak-services/src/chanserv.c
  1.15      +8 -0      zorak-services/src/clone.c
  1.20      +12 -1     zorak-services/src/conf.c
  1.2       +265 -21   zorak-services/src/dcc.c
  1.2       +14 -1     zorak-services/src/events.c
  1.8       +12 -0     zorak-services/src/irc_string.c
  1.39      +5 -5      zorak-services/src/main.c
  1.11      +13 -6     zorak-services/src/mem.c
  1.34      +154 -131  zorak-services/src/net.c
  1.6       +10 -10    zorak-services/src/nickserv.c
  1.18      +6 -0      zorak-services/src/oline.c
  1.26      +69 -338   zorak-services/src/operserv.c
  1.57      +18 -25    zorak-services/src/parse.c
  1.22      +4 -1      zorak-services/src/send.c
  1.32      +14 -14    zorak-services/src/users.c



wiz         2002/07/30 11:48:04 HST	(20020730_2)

  Modified files:
    src                  parse.c 
  Log:
  - Fixed a bug in m_error() that might be causing the SIGSEGV too
  
  Revision  Changes    Path
  1.56      +1 -1      zorak-services/src/parse.c



wiz         2002/07/30 11:31:16 HST	(20020730_1)

  Modified files:
    include              net.h 
    src                  Makefile.in main.c net.c 
  Added files:
    include              dcc.h events.h 
    src                  events.c 
  Log:
  - Replaced alarm() lameness with new events system, see add_events()
    Hopefully this will fix SIGSEGV on linux glibc with corrupt linked lists
  
  Revision  Changes    Path
  1.1       +28 -0     zorak-services/include/dcc.h (new)
  1.1       +37 -0     zorak-services/include/events.h (new)
  1.3       +2 -0      zorak-services/include/net.h
  1.22      +1 -0      zorak-services/src/Makefile.in

src/md5.c
/*
 * MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 *
 * Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
 * rights reserved.
 *
 * License to copy and use this software is granted provided that it
 * is identified as the "RSA Data Security, Inc. MD5 Message-Digest
 * Algorithm" in all material mentioning or referencing this software
 * or this function.
 *
 * License is also granted to make and use derivative works provided
 * that such works are identified as "derived from the RSA Data
 * Security, Inc. MD5 Message-Digest Algorithm" in all material
 * mentioning or referencing the derived work.
 *
 * RSA Data Security, Inc. makes no representations concerning either
 * the merchantability of this software or the suitability of this
 * software for any particular purpose. It is provided "as is"
 * without express or implied warranty of any kind.
 *
 * These notices must be retained in any copies of any part of this
 * documentation and/or software.
 *
 * This code is the same as the code published by RSA Inc.  It has been
 * edited for clarity and style only.
 */

#include "irc_types.h"
#include "irc_string.h"
#include "defines.h"
#include "md5.h"

WIZOSID("$Id: md5.c,v 1.5 2002/07/13 20:37:46 wiz Exp $");

typedef struct MD5Context {
  u_int32_t state[4];   /* state (ABCD) */
  u_int32_t count[2];   /* number of bits, modulo 2^64 (lsb first) */
  unsigned char buffer[64];     /* input buffer */
} MD5_CTX;

void   MD5Init (MD5_CTX *);
void   MD5Update (MD5_CTX *, const unsigned char *, unsigned int);
void   MD5Pad (MD5_CTX *);
void   MD5Final (unsigned char [16], MD5_CTX *);
char * MD5End(MD5_CTX *, char *);
char * MD5File(const char *, char *);
char * MD5Data(const unsigned char *, unsigned int, char *);

static void MD5Transform __P((u_int32_t [4], const unsigned char [64]));

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions. */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits. */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/*
 * FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
 * Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
	(a) += F ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}
#define GG(a, b, c, d, x, s, ac) { \
	(a) += G ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}
#define HH(a, b, c, d, x, s, ac) { \
	(a) += H ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}
#define II(a, b, c, d, x, s, ac) { \
	(a) += I ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
	(a) = ROTATE_LEFT ((a), (s)); \
	(a) += (b); \
	}

/* MD5 initialization. Begins an MD5 operation, writing a new context. */

void
MD5Init (context)
	MD5_CTX *context;
{

	context->count[0] = context->count[1] = 0;

	/* Load magic initialization constants.  */
	context->state[0] = 0x67452301;
	context->state[1] = 0xefcdab89;
	context->state[2] = 0x98badcfe;
	context->state[3] = 0x10325476;
}

/* 
 * MD5 block update operation. Continues an MD5 message-digest
 * operation, processing another message block, and updating the
 * context.
 */

void
MD5Update (context, input, inputLen)
	MD5_CTX *context;
	const unsigned char *input;
	unsigned int inputLen;
{
	unsigned int i, m_index, partLen;

	/* Compute number of bytes mod 64 */
	m_index = (unsigned int)((context->count[0] >> 3) & 0x3F);

	/* Update number of bits */
	if ((context->count[0] += ((u_int32_t)inputLen << 3))
	    < ((u_int32_t)inputLen << 3))
		context->count[1]++;
	context->count[1] += ((u_int32_t)inputLen >> 29);

	partLen = 64 - m_index;

	/* Transform as many times as possible. */
	if (inputLen >= partLen) {
		memcpy((void *)&context->buffer[m_index], (const void *)input,
		    partLen);
		MD5Transform (context->state, context->buffer);

		for (i = partLen; i + 63 < inputLen; i += 64)
			MD5Transform (context->state, &input[i]);

		m_index = 0;
	}
	else
		i = 0;

	/* Buffer remaining input */
	memcpy ((void *)&context->buffer[m_index], (const void *)&input[i],
	    inputLen-i);
}

/*
 * MD5 padding. Adds padding followed by original length.
 */

void
MD5Pad (context)
	MD5_CTX *context;
{
	unsigned char bits[8];
	unsigned int m_index, padLen;

	/* Save number of bits */
	memcpy (bits, context->count, 8);

	/* Pad out to 56 mod 64. */
	m_index = (unsigned int)((context->count[0] >> 3) & 0x3f);
	padLen = (m_index < 56) ? (56 - m_index) : (120 - m_index);
	MD5Update (context, PADDING, padLen);

	/* Append length (before padding) */
	MD5Update (context, bits, 8);
}

/*
 * MD5 finalization. Ends an MD5 message-digest operation, writing the
 * the message digest and zeroizing the context.
 */

void
MD5Final (digest, context)
	unsigned char digest[16];
	MD5_CTX *context;
{
	/* Do padding. */
	MD5Pad (context);

	/* Store state in digest */
	memcpy (digest, context->state, 16);

	/* Zeroize sensitive information. */
	memset ((void *)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block. */

static void
MD5Transform (state, block)
	u_int32_t state[4];
	const unsigned char block[64];
{
	u_int32_t a = state[0], b = state[1], c = state[2], d = state[3], x[16];

	memcpy (x, block, 64);

	/* Round 1 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
	FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
	FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
	FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
	FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
	FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
	FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
	FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
	FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
	FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
	FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
	FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
	FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
	FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
	FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
	FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
	FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

	/* Round 2 */
#define S21 5
#define S22 9
#define S23 14
#define S24 20
	GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
	GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
	GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
	GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
	GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
	GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
	GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
	GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
	GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
	GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
	GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
	GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
	GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
	GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
	GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
	GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

	/* Round 3 */
#define S31 4
#define S32 11
#define S33 16
#define S34 23
	HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
	HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
	HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
	HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
	HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
	HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
	HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
	HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
	HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
	HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
	HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
	HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
	HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
	HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
	HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
	HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

	/* Round 4 */
#define S41 6
#define S42 10
#define S43 15
#define S44 21
	II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
	II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
	II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
	II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
	II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
	II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
	II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
	II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
	II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
	II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
	II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
	II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
	II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
	II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
	II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
	II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;

	/* Zeroize sensitive information. */
	memset ((void *)x, 0, sizeof (x));
}

/* wiz code starts here */

#include "me.h"

char *
wiz_md5(char *stuff)
{
	MD5_CTX context;
	u_char digest[16];
	u_short i = 0, j = 0;

	MD5Init(&context);
	MD5Update(&context, (unsigned char *)stuff, strlen(stuff));
	MD5Final(digest, &context);

	me.md5buf[0] = '\0';
	for (i = 0, j = 0; i < 16; i++, j += 2)
		sprintf(&me.md5buf[j], "%02x", digest[i]);
	me.md5buf[32] = '\0';

	return me.md5buf;
}
